### PR TITLE
[SYCL][CUDA][libclc] Fixes atomics for below sm_60

### LINKED ISSUE #5008: [SYCL][CUDA] Fatal error: error in backend: Cannot select: intrinsic %llvm.nvvm.atomic.add.shared.i.cta
**Describe the bug**
After merging https://github.com/intel/llvm/pull/4820 compilation fails with error "Cannot select: intrinsic %llvm.nvvm.atomic.add.shared.i.cta".

```
fatal error: error in backend: Cannot select: intrinsic %llvm.nvvm.atomic.add.shared.i.cta
llvm-foreach:
clang++: error: clang frontend command failed with exit code 70 (use -v to see invocation)
clang version 14.0.0 (https://github.com/intel/llvm.git c855fd177a88d9fc6db87bc2a6c8001f0398ec9a)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: .../build/bin
clang++: note: diagnostic msg: Error generating preprocessed source(s).
```

**To Reproduce**

Let's compile the following example for the CUDA backend

```c++
#include <sycl/sycl.hpp>

int main() {
  constexpr static size_t buffer_size = 4;
  sycl::queue stream;
  sycl::buffer<int, 1> buffer(buffer_size);
  stream.submit(
    [&](sycl::handler &cgh) {
        sycl::accessor<int, 1, sycl::access_mode::read_write, sycl::access::target::local> acc(sycl::range<1>(buffer_size), cgh);
        auto acc_write = buffer.get_access<sycl::access::mode::write>(cgh);

        cgh.parallel_for<class demo_kernel>(buffer_size, [=](sycl::id<1> item) {
          sycl::atomic<int, sycl::access::address_space::local_space>(sycl::local_ptr<int>(acc.get_pointer())).fetch_add(1);         
        });
    });

  return 0;
}
```

using the following command:

```
clang++ -fsycl -fsycl-unnamed-lambda -fsycl-targets=nvptx64-nvidia-cuda -O2 example.cpp -o example  -lOpenCL
```

or with `-nocudalib` parameter, it's enough to see the error:

```
clang++ -fsycl -fsycl-unnamed-lambda -fsycl-targets=nvptx64-nvidia-cuda -O2 example.cpp -o example  -lOpenCL -nocudalib
```

The described above error (`fatal error: error in backend: Cannot select: intrinsic %llvm.nvvm.atomic.add.shared.i.cta`) occurs while the successful compilation is expected.

**Environment (please complete the following information):**

- OS: Ubuntu 20.04
- Target device and vendor: GPU: TITAN RTX
- DPC++ version:
https://github.com/intel/llvm/commit/c855fd177a88d9fc6db87bc2a6c8001f0398ec9a
- Dependencies version: 
NVIDIA-SMI 470.57.02
CUDA 11.4

**Additional context**

**Potential guilty commit**
commit 2ebde5f6011ea48f118219ae6fa8811f94e7acc7
Author: Tadej Ciglarič <tadej.ciglaric@codeplay.com>
Date:   Thu Nov 18 05:45:46 2021 +0100

[SYCL][CUDA][libclc] Added atomics with scopes and memory orders (#4820)

Added libclc implementations for CUDA atomics, including for various scopes and memory orders. They are implemented using LLVM intrinsics and exposed as clang builtins, which are than used to implement functions in libclc.

### PR DESCRIPTION
**Pull request by t4c1**
Fixes #5008. Fixes libclc implementation of CUDA atomics by using `__nvvm_reflect()` so as not to emit intrinsics that are unsupported on target sm version. If atomics are called with semantic order or scope that is not supported, `__builtin_trap()` is called instead.

### CODE REVIEW

**Review by bader — APPROVED** [review]

**Comment by bader** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/581

**Comment by bader** [comment]
`Basic/scalar_vec_access.cpp` and `Basic/stream/stream.cpp` tests from https://github.com/intel/llvm-test-suite/ are failing with this change:
```
[2021-11-24T18:52:01.806Z] ******************** TEST 'LLVM :: Basic/scalar_vec_access.cpp' FAILED ********************
[2021-11-24T18:52:01.806Z] Script:
[2021-11-24T18:52:01.806Z] --
[2021-11-24T18:52:01.806Z] : 'RUN: at line 1';    /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm.obj/bin/clang++      -fLLVM -fLLVM-targets=nvptx64-nvidia-cuda /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm-test-suite/LLVM/Basic/scalar_vec_access.cpp -o /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/Output/scalar_vec_access.cpp.tmp.out
[2021-11-24T18:52:01.806Z] : 'RUN: at line 2';   true /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/Output/scalar_vec_access.cpp.tmp.out
[2021-11-24T18:52:01.806Z] : 'RUN: at line 3';   true /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/Output/scalar_vec_access.cpp.tmp.out
[2021-11-24T18:52:01.806Z] : 'RUN: at line 4';    env LLVM_DEVICE_FILTER=cuda:gpu,host  /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/Output/scalar_vec_access.cpp.tmp.out | FileCheck /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm-test-suite/LLVM/Basic/scalar_vec_access.cpp
[2021-11-24T18:52:01.806Z] : 'RUN: at line 5';   true /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/Output/scalar_vec_access.cpp.tmp.out
[2021-11-24T18:52:01.806Z] --
[2021-11-24T18:52:01.806Z] Exit Code: 255
[2021-11-24T18:52:01.806Z] 
[2021-11-24T18:52:01.806Z] Command Output (stdout):
[2021-11-24T18:52:01.806Z] --
[2021-11-24T18:52:01.806Z] $ ":" "RUN: at line 1"
[2021-11-24T18:52:01.806Z] note: command had no output on stdout or stderr
[2021-11-24T18:52:01.806Z] $ "/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm.obj/bin/clang++" "-fLLVM" "-fLLVM-targets=nvptx64-nvidia-cuda" "/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm-test-suite/LLVM/Basic/scalar_vec_access.cpp" "-o" "/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/Output/scalar_vec_access.cpp.tmp.out"
[2021-11-24T18:52:01.806Z] # command stderr:
[2021-11-24T18:52:01.806Z] warning: linking module '/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm.obj/lib/clang/14.0.0/../../clc/remangled-l64-signed_char.libspirv-nvptx64--nvidiacl.bc': Linking two modules of different target triples: '/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm.obj/lib/clang/14.0.0/../../clc/remangled-l64-signed_char.libspirv-nvptx64--nvidiacl.bc' is 'nvptx64-unknown-nvidiacl' whereas '/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm-test-suite/LLVM/Basic/scalar_vec_access.cpp' is 'nvptx64-nvidia-cuda'
[2021-11-24T18:52:01.806Z]  [-Wlinker-warnings]
[2021-11-24T18:52:01.806Z] 1 warning generated.
[2021-11-24T18:52:01.806Z] ptxas fatal   : Unresolved extern function '__nvvm_reflect'
[2021-11-24T18:52:01.806Z] llvm-foreach: 
[2021-11-24T18:52:01.806Z] clang-14: error: ptxas command failed with exit code 255 (use -v to see invocation)
[2021-11-24T18:52:01.806Z] Target: x86_64-unknown-linux-gnu
[2021-11-24T18:52:01.806Z] Thread model: posix
[2021-11-24T18:52:01.806Z] InstalledDir: /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm.obj/bin
[2021-11-24T18:52:01.806Z] clang-14: note: diagnostic msg: Error generating preprocessed source(s).
[2021-11-24T18:52:01.806Z] 
[2021-11-24T18:52:01.806Z] error: command failed with exit status: 255
[2021-11-24T18:52:01.806Z] 
[2021-11-24T18:52:01.806Z] --
[2021-11-24T18:52:01.806Z] 
[2021-11-24T18:52:01.806Z] ********************
```

```
[2021-11-24T18:52:12.821Z] ******************** TEST 'LLVM :: Basic/stream/stream.cpp' FAILED ********************
[2021-11-24T18:52:12.821Z] Script:
[2021-11-24T18:52:12.821Z] --
[2021-11-24T18:52:12.821Z] : 'RUN: at line 1';    /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm.obj/bin/clang++      -fLLVM -fLLVM-targets=nvptx64-nvidia-cuda /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm-test-suite/LLVM/Basic/stream/stream.cpp -o /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/stream/Output/stream.cpp.tmp.out
[2021-11-24T18:52:12.821Z] : 'RUN: at line 4';   true /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/stream/Output/stream.cpp.tmp.out
[2021-11-24T18:52:12.821Z] : 'RUN: at line 5';   env LLVM_DEVICE_FILTER=cuda:gpu,host  /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/stream/Output/stream.cpp.tmp.out | FileCheck /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm-test-suite/LLVM/Basic/stream/stream.cpp
[2021-11-24T18:52:12.821Z] : 'RUN: at line 6';   true /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/stream/Output/stream.cpp.tmp.out
[2021-11-24T18:52:12.821Z] --
[2021-11-24T18:52:12.821Z] Exit Code: 255
[2021-11-24T18:52:12.821Z] 
[2021-11-24T18:52:12.821Z] Command Output (stdout):
[2021-11-24T18:52:12.821Z] --
[2021-11-24T18:52:12.821Z] $ ":" "RUN: at line 1"
[2021-11-24T18:52:12.821Z] note: command had no output on stdout or stderr
[2021-11-24T18:52:12.821Z] $ "/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm.obj/bin/clang++" "-fLLVM" "-fLLVM-targets=nvptx64-nvidia-cuda" "/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm-test-suite/LLVM/Basic/stream/stream.cpp" "-o" "/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/build/LLVM/Basic/stream/Output/stream.cpp.tmp.out"
[2021-11-24T18:52:12.821Z] # command stderr:
[2021-11-24T18:52:12.821Z] warning: linking module '/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm.obj/lib/clang/14.0.0/../../clc/remangled-l64-signed_char.libspirv-nvptx64--nvidiacl.bc': Linking two modules of different target triples: '/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm.obj/lib/clang/14.0.0/../../clc/remangled-l64-signed_char.libspirv-nvptx64--nvidiacl.bc' is 'nvptx64-unknown-nvidiacl' whereas '/localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm-test-suite/LLVM/Basic/stream/stream.cpp' is 'nvptx64-nvidia-cuda'
[2021-11-24T18:52:12.821Z]  [-Wlinker-warnings]
[2021-11-24T18:52:12.821Z] 1 warning generated.
[2021-11-24T18:52:12.821Z] ptxas fatal   : Unresolved extern function '__nvvm_reflect'
[2021-11-24T18:52:12.821Z] llvm-foreach: 
[2021-11-24T18:52:12.821Z] clang-14: error: ptxas command failed with exit code 255 (use -v to see invocation)
[2021-11-24T18:52:12.821Z] Target: x86_64-unknown-linux-gnu
[2021-11-24T18:52:12.821Z] Thread model: posix
[2021-11-24T18:52:12.821Z] InstalledDir: /localdisk2/iusers/nstester/Codeplay/workspace/LLVM_CI/intel/Lin/LLVM_Test_Suite_CUDA/llvm.obj/bin
[2021-11-24T18:52:12.821Z] clang-14: note: diagnostic msg: Error generating preprocessed source(s).
[2021-11-24T18:52:12.821Z] 
[2021-11-24T18:52:12.821Z] error: command failed with exit status: 255
[2021-11-24T18:52:12.821Z] 
[2021-11-24T18:52:12.821Z] --
[2021-11-24T18:52:12.821Z] 
[2021-11-24T18:52:12.821Z] ********************
```

It looks like we need to run `NVVMReflect` pass to resolve that issue. See https://llvm.org/docs/NVPTXUsage.html#common-issues.

**Review by bader — CHANGES_REQUESTED** [review]
Please, fix `ptxas fatal   : Unresolved extern function '__nvvm_reflect'` issue.

File: libclc/ptx-nvidiacl/libspirv/group/collectives.cl [hunk 1]
```diff
@@ -191,7 +191,7 @@ __clc__SubgroupBitwiseAny(uint op, bool predicate, bool *carry) {
   _CLC_DEF _CLC_OVERLOAD _CLC_CONVERGENT TYPE __CLC_APPEND(                    \
       __clc__Subgroup, NAME)(uint op, TYPE x, TYPE * carry) {                  \
     /* Fast path for warp reductions for sm_80+ */                             \
-    if (__nvvm_reflect("__CUDA_ARCH") >= 800 && op == Reduce) {                \
+    if (__clc_nvvm_reflect_arch() >= 800 && op == Reduce) {                      \
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/group/collectives.cl:194]:
```suggestion
    if (__clc_nvvm_reflect_arch() >= 800 && op == Reduce) {                    \
```

**Comment by t4c1** [comment]
So it looks like calling `__nvvm_reflect` from OpenCL is a bit broken. `__nvvm_reflect` requires an argument of type `char*`, while in OpenCL strings are in `__constant` AS. So we were declaring `__nvvm_reflect(__constant char*)`, which works for uses in libclc, but breaks the uses of reflect in libdevice. I fixed this by declaring the string for reflect in IR instead.

**Comment by bader** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/590

**Review by bader — APPROVED** [review]

**Pull request merged**